README.md
# init-yyl-config

用于初始 yyl.config.js

## 安装

```bash
yarn add init-yyl-config --dev
```

## 使用

```javascript
// yyl.config.js
const { initYylConfig } = require('init-yyl-config')
return initYylConfig()
```

## 文档

[这里](https://htmlpreview.github.io/?https://raw.githubusercontent.com/jackness1208/init-yyl-config/master/docs/modules.html)


history.md
# 历史版本

## 0.2.1 (2021-05-28)

- fix: 修复配置 yylConfig.alias 不生效问题

## 0.2.0 (2021-05-28)

- feat: 采用更人性的调用方式

## 0.1.4 (2021-05-28)

- feat: 更新 `yyl-config-types@0.5.4`

## 0.1.3 (2021-04-13)

- feat: 调整 types

## 0.1.2 (2021-04-13)

- feat: 更新 `yyl-config-types@0.4.0`


test/jest/index.test.js
const { TestScheduler } = require('jest')
const { initYylConfig } = require('../../')

test('usage test', () => {
  const rFn = initYylConfig(() => {
    return {
      name: 'projectname',
      workflow: 'webpack',
      platform: 'pc',
      version: '1.0.0',
      yarn: true
    }
  })
  expect(rFn({ env: {} })).toEqual({
    name: 'projectname',
    workflow: 'webpack',
    platform: 'pc',
    version: '1.0.0',
    yarn: true,
    alias: {
      basePath: '/',
      cssDest: 'dist/css',
      destRoot: './dist',
      dirname: './',
      htmlDest: 'dist/html',
      imagesDest: 'dist/image',
      jsDest: 'dist/js',
      revDest: 'dist/assets',
      revRoot: 'dist/',
      root: 'dist/',
      srcRoot: './src'
    },
    commit: {
      hostname: '/',
      revAddr: '/assets'
    },
    dest: {
      basePath: '/',
      cssPath: 'css',
      htmlPath: 'html',
      imagesPath: 'image',
      jsPath: 'js',
      revPath: 'assets'
    },
    localserver: {
      port: 5000,
      root: './dist'
    },
    proxy: {
      localRemote: {},
      port: 8887
    },
    px2rem: false,
    seed: 'base'
  })
})


test/case/base/yyl.config.ts
import { initYylConfig, NiceYylConfig } from '../../../'
import pkg from './package.json'

// + vars
const PROJECT_NAME = pkg.name
const WORKFLOW = 'webpack'
const PLATFORM = 'pc'
const VERSION = '4.0.0'
const SRC_ROOT = './src'
const USE_YARN = true
// - vars

const hostname = 'www.yourhost.com'
const staticHost = 'wwww.yourhost.com'
module.exports = initYylConfig(({ env }) => {
  const yylConfig: NiceYylConfig = {
    name: PROJECT_NAME,
    workflow: WORKFLOW,
    platform: PLATFORM,
    version: VERSION,
    yarn: USE_YARN,
    dest: {
      basePath: `/project/${PROJECT_NAME}/${PLATFORM}`,
      jsPath: 'js',
      cssPath: 'css',
      htmlPath: 'html',
      imagesPath: 'images',
      revPath: 'asset'
    },
    proxy: {
      homePage: `http://${hostname}/web/${PROJECT_NAME}/`,
      localRemote: (() => {
        const r = {}
        r[
          `http://${hostname}/web/${PROJECT_NAME}`
        ] = `http://127.0.0.1:5000/project/${PROJECT_NAME}/${PLATFORM}/html`
        return r
      })()
    },
    alias: {
      srcRoot: SRC_ROOT
    },
    commit: {
      hostname: `//${staticHost}`,
      revAddr: `http://${staticHost}/project/${PROJECT_NAME}/${PLATFORM}/assets/rev-manifest.json`
    }
  }

  return yylConfig
})


output/index.js
/*!
 * init-yyl-config cjs 0.2.1
 * (c) 2020 - 2021 
 * Released under the MIT License.
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var path = require('path');
var yylConfigTypes = require('yyl-config-types');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var path__default = /*#__PURE__*/_interopDefaultLegacy(path);

/** 默认项目配置 */
const DEFAULT_PROJECT_INFO = {
    /** 项目名称 */
    name: '',
    /** seed 类型 */
    workflow: 'webpack',
    /** 平台类型 */
    platform: 'pc',
    /** yyl 版本 */
    version: '',
    /** 是否使用 yarn */
    yarn: true,
    seed: 'base'
};
/** 默认服务器配置 */
const DEFAULT_LOCAL_SEVER_CONFIG = {
    /** server映射路径 */
    root: './dist',
    /** 端口 */
    port: 5000
};
/** 默认反向代理配置 */
const DEFAULT_PROXY_CONFIG = {
    /** 端口 */
    port: 8887,
    /** 映射 map */
    localRemote: {}
};
/** 默认 输出 配置 */
const DEFAULT_DEST_CONFIG = {
    basePath: '/',
    jsPath: 'js',
    cssPath: 'css',
    htmlPath: 'html',
    imagesPath: 'image',
    revPath: 'assets'
};
const DEFAULT_COMMIT_CONFIG = {
    revAddr: '',
    hostname: '/'
};
/** 初始化 yyl.config */
function initYylConfig(fn) {
    const r = (props) => {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const { env } = props;
        let yylConfig = fn({ env });
        // 基本信息初始化
        yylConfig = Object.assign(Object.assign({}, DEFAULT_PROJECT_INFO), yylConfig);
        // 本地 server
        yylConfig.localserver = Object.assign(Object.assign({}, DEFAULT_LOCAL_SEVER_CONFIG), yylConfig.localserver);
        // proxy
        yylConfig.proxy = Object.assign(Object.assign({}, DEFAULT_PROXY_CONFIG), yylConfig.proxy);
        // dest
        yylConfig.dest = Object.assign(Object.assign({}, DEFAULT_DEST_CONFIG), yylConfig.dest);
        // commit
        yylConfig.commit = Object.assign(Object.assign({}, DEFAULT_COMMIT_CONFIG), yylConfig.commit);
        if (!yylConfig.commit.revAddr) {
            yylConfig.commit.revAddr = `${(_a = yylConfig.commit) === null || _a === void 0 ? void 0 : _a.hostname}${yylConfig.dest.basePath}/${yylConfig.dest.revPath}`.replace(/[\/]+/, '/');
        }
        // px2rem
        if (yylConfig.px2rem === undefined) {
            yylConfig.px2rem = yylConfig.platform === 'mobile';
        }
        // alias 初始化
        const DEST_BASE_PATH = path__default['default'].join(yylConfig.localserver.root || process.cwd(), yylConfig.dest.basePath);
        yylConfig.alias = Object.assign({ 
            /** 输出目录中 到 html, js, css, image 层 的路径 */
            root: DEST_BASE_PATH, 
            /** rev 输出内容的相对地址 */
            revRoot: DEST_BASE_PATH, 
            /** dest 地址 */
            destRoot: (_b = yylConfig.localserver) === null || _b === void 0 ? void 0 : _b.root, 
            /** src 地址 */
            srcRoot: './src', 
            /** 项目根目录 */
            dirname: './', 
            /** js 输出地址 */
            jsDest: path__default['default'].join(DEST_BASE_PATH, (_c = yylConfig.dest) === null || _c === void 0 ? void 0 : _c.jsPath), 
            /** html 输出地址 */
            htmlDest: path__default['default'].join(DEST_BASE_PATH, (_d = yylConfig.dest) === null || _d === void 0 ? void 0 : _d.htmlPath), 
            /** css 输出地址 */
            cssDest: path__default['default'].join(DEST_BASE_PATH, (_e = yylConfig.dest) === null || _e === void 0 ? void 0 : _e.cssPath), 
            /** images 输出地址 */
            imagesDest: path__default['default'].join(DEST_BASE_PATH, (_f = yylConfig.dest) === null || _f === void 0 ? void 0 : _f.imagesPath), 
            /** rev-manifest 输出地址 */
            revDest: path__default['default'].join(DEST_BASE_PATH, (_g = yylConfig.dest) === null || _g === void 0 ? void 0 : _g.revPath), basePath: (_h = yylConfig.dest) === null || _h === void 0 ? void 0 : _h.basePath }, yylConfig.alias);
        // 初始化处理
        return yylConfig;
    };
    return r;
}

exports.DEFAULT_COMMIT_CONFIG = DEFAULT_COMMIT_CONFIG;
exports.DEFAULT_DEST_CONFIG = DEFAULT_DEST_CONFIG;
exports.DEFAULT_LOCAL_SEVER_CONFIG = DEFAULT_LOCAL_SEVER_CONFIG;
exports.DEFAULT_PROJECT_INFO = DEFAULT_PROJECT_INFO;
exports.DEFAULT_PROXY_CONFIG = DEFAULT_PROXY_CONFIG;
exports.initYylConfig = initYylConfig;
Object.keys(yylConfigTypes).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return yylConfigTypes[k];
    }
  });
});


output/index.d.ts
import { YylConfig, Env, DestConfig, LocalserverConfig, ProxyConfig, CommitConfig, YylConfigEntryFn } from 'yyl-config-types';
export * from 'yyl-config-types';
export interface InitYylConfigFunctionProps {
    env: Env;
}
export interface NiceYylConfig extends Omit<YylConfig, 'commit' | 'alias'> {
    alias?: {
        [key: string]: string;
    };
    commit: {
        revAddr?: string;
        hostname: string;
        staticHost?: string;
        mainHost?: string;
    };
}
export declare type InitYylConfigCallback = (props: InitYylConfigFunctionProps) => NiceYylConfig;
/** 默认项目配置 */
export declare const DEFAULT_PROJECT_INFO: Required<Pick<YylConfig, 'name' | 'workflow' | 'platform' | 'version' | 'yarn' | 'seed'>>;
/** 默认服务器配置 */
export declare const DEFAULT_LOCAL_SEVER_CONFIG: LocalserverConfig;
/** 默认反向代理配置 */
export declare const DEFAULT_PROXY_CONFIG: ProxyConfig;
/** 默认 输出 配置 */
export declare const DEFAULT_DEST_CONFIG: DestConfig;
export declare const DEFAULT_COMMIT_CONFIG: CommitConfig;
/** 初始化 yyl.config */
export declare function initYylConfig(fn: InitYylConfigCallback): YylConfigEntryFn;


src/index.ts
import path from 'path'
import {
  YylConfig,
  Env,
  DestConfig,
  LocalserverConfig,
  ProxyConfig,
  CommitConfig,
  YylConfigEntryFn
} from 'yyl-config-types'

export * from 'yyl-config-types'

export interface InitYylConfigFunctionProps {
  env: Env
}
export interface NiceYylConfig extends Omit<YylConfig, 'commit' | 'alias'> {
  alias?: {
    [key: string]: string
  }
  commit: {
    revAddr?: string
    hostname: string
    staticHost?: string
    mainHost?: string
  }
}

export type InitYylConfigCallback = (props: InitYylConfigFunctionProps) => NiceYylConfig

/** 默认项目配置 */
export const DEFAULT_PROJECT_INFO: Required<
  Pick<YylConfig, 'name' | 'workflow' | 'platform' | 'version' | 'yarn' | 'seed'>
> = {
  /** 项目名称 */
  name: '',
  /** seed 类型 */
  workflow: 'webpack',
  /** 平台类型 */
  platform: 'pc',
  /** yyl 版本 */
  version: '',
  /** 是否使用 yarn */
  yarn: true,
  seed: 'base'
}

/** 默认服务器配置 */
export const DEFAULT_LOCAL_SEVER_CONFIG: LocalserverConfig = {
  /** server映射路径 */
  root: './dist',
  /** 端口 */
  port: 5000
}

/** 默认反向代理配置 */
export const DEFAULT_PROXY_CONFIG: ProxyConfig = {
  /** 端口 */
  port: 8887,
  /** 映射 map */
  localRemote: {}
}

/** 默认 输出 配置 */
export const DEFAULT_DEST_CONFIG: DestConfig = {
  basePath: '/',
  jsPath: 'js',
  cssPath: 'css',
  htmlPath: 'html',
  imagesPath: 'image',
  revPath: 'assets'
}

export const DEFAULT_COMMIT_CONFIG: CommitConfig = {
  revAddr: '',
  hostname: '/'
}

/** 初始化 yyl.config */
export function initYylConfig(fn: InitYylConfigCallback): YylConfigEntryFn {
  const r: YylConfigEntryFn = (props) => {
    const { env } = props
    let yylConfig = fn({ env })

    // 基本信息初始化
    yylConfig = {
      ...DEFAULT_PROJECT_INFO,
      ...yylConfig
    }

    // 本地 server
    yylConfig.localserver = {
      ...DEFAULT_LOCAL_SEVER_CONFIG,
      ...yylConfig.localserver
    }

    // proxy
    yylConfig.proxy = {
      ...DEFAULT_PROXY_CONFIG,
      ...yylConfig.proxy
    }

    // dest
    yylConfig.dest = {
      ...DEFAULT_DEST_CONFIG,
      ...yylConfig.dest
    }

    // commit
    yylConfig.commit = {
      ...DEFAULT_COMMIT_CONFIG,
      ...yylConfig.commit
    }
    if (!yylConfig.commit.revAddr) {
      yylConfig.commit.revAddr = `${yylConfig.commit?.hostname}${yylConfig.dest.basePath}/${yylConfig.dest.revPath}`.replace(
        /[\/]+/,
        '/'
      )
    }

    // px2rem
    if (yylConfig.px2rem === undefined) {
      yylConfig.px2rem = yylConfig.platform === 'mobile'
    }

    // alias 初始化
    const DEST_BASE_PATH = path.join(
      yylConfig.localserver.root || process.cwd(),
      yylConfig.dest.basePath as string
    )
    yylConfig.alias = {
      /** 输出目录中 到 html, js, css, image 层 的路径 */
      root: DEST_BASE_PATH,
      /** rev 输出内容的相对地址 */
      revRoot: DEST_BASE_PATH,
      /** dest 地址 */
      destRoot: yylConfig.localserver?.root as string,
      /** src 地址 */
      srcRoot: './src',
      /** 项目根目录 */
      dirname: './',
      /** js 输出地址 */
      jsDest: path.join(DEST_BASE_PATH, yylConfig.dest?.jsPath as string),
      /** html 输出地址 */
      htmlDest: path.join(DEST_BASE_PATH, yylConfig.dest?.htmlPath as string),
      /** css 输出地址 */
      cssDest: path.join(DEST_BASE_PATH, yylConfig.dest?.cssPath as string),
      /** images 输出地址 */
      imagesDest: path.join(DEST_BASE_PATH, yylConfig.dest?.imagesPath as string),
      /** rev-manifest 输出地址 */
      revDest: path.join(DEST_BASE_PATH, yylConfig.dest?.revPath as string),
      basePath: yylConfig.dest?.basePath as string,
      ...yylConfig.alias
    }

    // 初始化处理
    return yylConfig as YylConfig
  }
  return r
}
